tests/test_coordinated_workers/test_coordinator.py

```diff
@@ -281,3 +284,103 @@ def test_tracing_receivers_urls(
             "otlp_http": "5.6.7.8:4318",
             "otlp_grpc": "5.6.7.8:4317",
         }
+
+
+@pytest.fixture
+def cluster_state():
+    requires_relations = {
+        endpoint: testing.Relation(endpoint=endpoint, interface=interface["interface"])
+        for endpoint, interface in {
+            "my-certificates": {"interface": "certificates"},
+            "my-logging": {"interface": "loki_push_api"},
+            "my-charm-tracing": {"interface": "tracing"},
+            "my-workload-tracing": {"interface": "tracing"},
+        }.items()
+    }
+    requires_relations["my-s3"] = testing.Relation(
+        "my-s3",
+        interface="s3",
+        remote_app_data={
+            "endpoint": "s3",
+            "bucket": "foo-bucket",
+            "access-key": "my-access-key",
+            "secret-key": "my-secret-key",
+        },
+    )
+    requires_relations["cluster_worker0"] = testing.Relation(
+        "my-cluster",
+        remote_app_name="worker0",
+        remote_app_data=ClusterRequirerAppData(role="read").dump(),
+    )
+    requires_relations["cluster_worker1"] = testing.Relation(
+        "my-cluster",
+        remote_app_name="worker1",
+        remote_app_data=ClusterRequirerAppData(role="write").dump(),
+    )
+    requires_relations["cluster_worker2"] = testing.Relation(
+        "my-cluster",
+        remote_app_name="worker2",
+        # remote_app_data=ClusterRequirerAppData(role="backend").dump(),
+    )
+
+    provides_relations = {
+        endpoint: testing.Relation(endpoint=endpoint, interface=interface["interface"])
+        for endpoint, interface in {
+            "my-dashboards": {"interface": "grafana_dashboard"},
+            "my-metrics": {"interface": "prometheus_scrape"},
+        }.items()
+    }
+
+    return testing.State(
+        containers={
+            testing.Container("nginx", can_connect=True),
+            testing.Container("nginx-prometheus-exporter", can_connect=True),
+        },
+        relations=list(requires_relations.values()) + list(provides_relations.values()),
+    )
+
+
+@pytest.fixture()
+def cluster_charm(request):
```

Comment:
why can't you reuse the `coordinator_charm` fixture instead of adding this very similar one?
also, I don't think you need the `request` param

Reply:
I can, I have switch to using it.